src/source/position.rs

```diff
@@ -0,0 +1,133 @@
+use std::time::Duration;
+
+use crate::{Sample, Source};
+
+use super::SeekError;
+
+/// Internal function that builds a `TrackPosition` object.
+pub fn trackable<I>(source: I) -> TrackPosition<I> {
+    TrackPosition {
+        input: source,
+        samples_elapsed: 0,
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct TrackPosition<I> {
+    input: I,
+    samples_elapsed: usize,
+}
+
+impl<I> TrackPosition<I> {
+    /// Returns a reference to the inner source.
+    #[inline]
+    pub fn inner(&self) -> &I {
+        &self.input
+    }
+
+    /// Returns a mutable reference to the inner source.
+    #[inline]
+    pub fn inner_mut(&mut self) -> &mut I {
+        &mut self.input
+    }
+
+    /// Returns the inner source.
+    #[inline]
+    pub fn into_inner(self) -> I {
+        self.input
+    }
+}
+
+impl<I> TrackPosition<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    /// Returns the inner source.
+    #[inline]
+    pub fn get_pos(&self) -> f64 {
+        self.samples_elapsed as f64 / self.input.sample_rate() as f64 / self.input.channels() as f64
+    }
+}
+
+impl<I> Iterator for TrackPosition<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    type Item = I::Item;
+
+    #[inline]
+    fn next(&mut self) -> Option<I::Item> {
+        let item = self.input.next();
+        if item.is_some() {
+            self.samples_elapsed += 1;
+        };
+        item
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.input.size_hint()
+    }
+}
+
+impl<I> Source for TrackPosition<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    #[inline]
+    fn current_frame_len(&self) -> Option<usize> {
+        self.input.current_frame_len()
```

Comment:
after `current_frame_len` the sample rate may change. 
One way to deal with that is to:
- collect samples until the frame rate changes 
- convert the collected samples using the old frame-rate and add them to a duration
- change `get_pos()` to return the newly collected samples * new frame rate + duration

Feel free to experiment with other approaches to find what fits best. Performance matters but as always _premature optimization is the root of all evil_. An if statement that fires every `frame_len` in `next()` might very well pulled up by the optimizer and merged with the iterator in some more optimal fashion. Better to measure the impact before spending time on optimizing.

Reply:
Do you happen to know which decoder currently handle this correctly? I've tried with FLAC using https://github.com/ietf-wg-cellar/flac-test-files/blob/main/uncommon/01%20-%20changing%20samplerate.flac with the symphonia decoder and after switching to the next sample rate it stops working and [claxon](https://github.com/ruuda/claxon) doesn't stop playing but doesn't seem to record the new sample rate (just reports the 32 kHz all the way trough) and it ends up 'pitching' the audio. (FWIW, VLC also borks on this only mpv seems to somewhat handle it but has incorrect position after switching and wrong total duration). Shouldn't be hard to handle these cases correctly, but at least the FLAC decoder is stopping me from testing this and I don't know of any other format that supports variable sample rate 😅

Reply:
nice! thanks a lot for that :+1: I had no idea these where broken (rodio's test suit is kinda sparse atm)

Reply:
Once the decoders are fixed https://github.com/RustAudio/rodio/pull/585/commits/0e6d5a1da2d75d462f145dcd4ad54d5821599deb should make the position tracking work correctly with. Doing a few local tests doesn't hint at performance problems (but only a profiler would be able to confirm that) and tracking the position of https://github.com/ietf-wg-cellar/flac-test-files/blob/main/uncommon/01%20-%20changing%20samplerate.flac works flawlessly.